### ISSUE TITLE
False positive: "Condition always true" for `(keys m)` / `(vals m)`

### ISSUE THREAD

**Issue by smithtim** [clj-kondo/clj-kondo#2928]
- [x] I have read the [Clojure etiquette](https://clojure.org/community/etiquette) and will respect it when communicating on this platform.

**version**

clj-kondo v2026.07.24

**macro usage**

Not related to macro usage.

**platform**

Reproduces on both the native binary and the JVM version.

**editor**

N/A — reproduced directly from the command line.

**problem**

clj-kondo reports `Condition always true` when the result of `(keys m)` (or
`(vals m)`) is used as a condition (`if`, `if-let`, `when`, ...). But `(keys {})`
returns `nil` for an empty map, so the condition is **not** always true — the
else branch is reachable whenever the map is empty. An empty map is an ordinary
value, so guarding on `(keys m)` is legitimate and should not be flagged.

The result type of `keys`/`vals` is defined as the non-nilable `:seq` in
`src/clj_kondo/impl/types/clojure/core.clj`, so the type-checker treats it as
always-truthy. Since `(keys {})`/`(vals {})` are `nil`, the return type should be
nilable.

**repro**

```clojure
(ns repro)

(defn f [m]
  (if-let [ks (keys m)]
    ks
    :empty-map))
```

```
$ clj-kondo --lint repro.clj
repro.clj:4:15: warning: Condition always true
linting took 28ms, errors: 0, warnings: 1
```

`vals` behaves identically. Runtime confirms the else branch is reachable:

```clojure
(keys {})                                              ;=> nil   (falsy)
((fn [m] (if-let [ks (keys m)] ks :empty-map)) {})     ;=> :empty-map   (else branch)
((fn [m] (if-let [ks (keys m)] ks :empty-map)) {:a 1}) ;=> (:a)
```

**config**

Not related to `.clj-kondo/config.edn`. Reproduces with no configuration present.

**expected behavior**

No warning for `(keys m)` / `(vals m)` used as a condition, since the result is
nilable (`nil` for empty collections).

<details>
<summary>Additional context: root cause and when it started</summary>

The `keys`/`vals` return type was changed from `:seqable` to the non-nilable
`:seq` in `72d1914d` (2020-01-31, PR #725, "Change more seqables return types
into seq"), which did not account for `keys`/`vals` returning `nil` on empty
collections. This stayed latent for years because function-call conditions were
not judged.

The warning first appears in PR #2902 "Constant condition" (`0d7cfa5c`,
2026-07-20), which added `lint-deferred-conditions!` — this resolves the type of
conditions that are function *calls* (like `(keys m)`) and reports
`Condition always true` when that type is non-nilable, surfacing the latent
inaccurate type. Bisected by building from source:

| commit | result |
|---|---|
| `111b2bc0` (parent of #2902) | no warning |
| `0d7cfa5c` (#2902) | `Condition always true` |
| `v2026.01.19` (release before #2902) | no warning |
| `v2026.07.24` (release after #2902) | `Condition always true` |

Making `keys`/`vals` return a nilable type should resolve the false positive.
</details>

**Comment by borkdude** [comment]
The docstring of `keys` says that it returns a sequence of keys. The `nil` behavior seems like undefined behavior to me?

cc @puredanger

**Comment by borkdude** [comment]
@puredanger replied:

> I think having looked at the code, that right now you could and should be able to rely on this. but if you want a stronger guarantee, you should call seq on it

So I guess widening the return type to seq + nil would be good.

**Comment by puredanger** [comment]
Certainly, the impl is going to a lot of trouble to intentionally return nil for empty maps and imo nil **is** a sequence, so I would definitely expect `nil` to be returned from `keys` or `vals`.
